### PR TITLE
build container image.

### PR DESCRIPTION
**Pull request by antonio-pedro99**

### CODE REVIEW

**Review by kyguy — APPROVED** [review]
Solid work @antonio-pedro99!

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+# Set JAVA_HOME env var
+ENV JAVA_HOME /usr/lib/jvm/jre-${JAVA_VERSION}
+
+# Add strimzi user with UID 1001
+# The user is in the group 0 to have access to the mounted volumes and storage
+RUN useradd -r -m -u 1001 -g 0 strimzi
+
+ARG STRIMZI_MQTT_BRIDGE_VERSION=1.0-SNAPSHOT
+ENV STRIMZI_MQTT_BRIDGE_VERSION ${STRIMZI_MQTT_BRIDGE_VERSION}
```
Comment (ppatierno) [inline Dockerfile:19]:
```suggestion
ENV STRIMZI_MQTT_BRIDGE_VERSION ${strimzi_mqtt_bridge_version}
```

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+ENV STRIMZI_HOME=/opt/strimzi
+RUN mkdir -p ${STRIMZI_HOME}
+WORKDIR ${STRIMZI_HOME}
+
+COPY target/mqtt-bridge-${STRIMZI_MQTT_BRIDGE_VERSION}/mqtt-bridge-${STRIMZI_MQTT_BRIDGE_VERSION} ./
```
Comment (ppatierno) [inline Dockerfile:24]:
```suggestion
COPY target/mqtt-bridge-${strimzi_mqtt_bridge_version}/mqtt-bridge-${strimzi_mqtt_bridge_version} ./
```

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,28 @@
+FROM registry.access.redhat.com/ubi8/ubi-minimal:latest
+ARG JAVA_VERSION=17
+ARG TARGETPLATFORM
```
Comment (ppatierno) [inline Dockerfile:3]: it's not used here but I would use it by adding the section about installing `tini` as it's done for the Strimzi HTTP bridge.

File: Makefile.docker [hunk 4]
```diff
@@ -0,0 +1,34 @@
+TOPDIR=$(dir $(lastword $(MAKEFILE_LIST)))
```
Comment (ppatierno) [inline Makefile.docker:1]:
```suggestion
# Makefile.docker contains the shared tasks for building, tagging and pushing Docker images.
# This file is included into the Makefile files which contain the Dockerfile files (E.g.
# kafka-base, kafka etc.).
#
# The DOCKER_ORG (default is name of the current user) and DOCKER_TAG (based on Git Tag,
# default latest) variables are used to name the Docker image. DOCKER_REGISTRY identifies
# the registry where the image will be pushed (default is Docker Hub).
TOPDIR=$(dir $(lastword $(MAKEFILE_LIST)))
```

I would leave the comments block we have in the Strimzi HTTP bridge to clarify what's defined here.

File: Makefile.docker [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+DOCKERFILE_DIR     ?= ./
+DOCKER_REGISTRY    ?= quay.io
```
Comment (ppatierno) [inline Makefile.docker:3]:
```suggestion
DOCKER_REGISTRY    ?= docker.io
```

Prefer having Docker as default for community users (as in the Strimzi HTTP bridge) even if Strimzi builds use quay.io in the end.

File: Makefile.docker [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+endif
+
+.PHONY: docker_build
+docker_build:
+	# Build Docker image ...
+	docker $(DOCKER_BUILDX) build $(DOCKER_PLATFORM) $(DOCKER_BUILD_ARGS) --build-arg strimzi_mqtt_bridge_version=$(RELEASE_VERSION) -t strimzi/$(PROJECT_NAME):latest $(DOCKERFILE_DIR)
+	# Also tag with $(BUILD_TAG)
+	docker tag strimzi/$(PROJECT_NAME):latest strimzi/$(PROJECT_NAME):$(BUILD_TAG)$(DOCKER_PLATFORM_TAG_SUFFIX)
+
+.PHONY: docker_load
```
Comment (ppatierno) [inline Makefile.docker:21]:
```suggestion
.PHONY: docker_save
docker_save:
	# Saves the container as TGZ file
	docker save strimzi/$(PROJECT_NAME):$(BUILD_TAG)$(DOCKER_PLATFORM_TAG_SUFFIX) | gzip > mqtt-bridge$(DOCKER_PLATFORM_TAG_SUFFIX).tar.gz

.PHONY: docker_load
```

Added the `docker_save` target which is useful to the image as tar.gz to inspect with tools or load offline (as you have with the `docker_load` target below)

File: Makefile.docker [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+.PHONY: docker_tag
+docker_tag:
+	# Tag the $(BUILD_TAG) image we built with the given $(DOCKER_TAG) tag
+	docker tag strimzi/$(PROJECT_NAME):$(BUILD_TAG)$(DOCKER_PLATFORM_TAG_SUFFIX) $(DOCKER_REGISTRY)/$(DOCKER_ORG)/$(PROJECT_NAME):$(DOCKER_TAG)$(DOCKER_PLATFORM_TAG_SUFFIX)
+
+.PHONY: docker_push
+docker_push: docker_tag
+	# Push the $(DOCKER_TAG)-tagged image to the registry
+	docker push $(DOCKER_REGISTRY)/$(DOCKER_ORG)/$(PROJECT_NAME):$(DOCKER_TAG)$(DOCKER_PLATFORM_TAG_SUFFIX)
```
Comment (ppatierno) [inline Makefile.docker:34]:
```suggestion
	docker push $(DOCKER_REGISTRY)/$(DOCKER_ORG)/$(PROJECT_NAME):$(DOCKER_TAG)$(DOCKER_PLATFORM_TAG_SUFFIX)
	
.PHONY: docker_amend_manifest
docker_amend_manifest:
	# Create / Amend the manifest
	docker manifest create $(DOCKER_REGISTRY)/$(DOCKER_ORG)/$(PROJECT_NAME):$(DOCKER_TAG) --amend $(DOCKER_REGISTRY)/$(DOCKER_ORG)/$(PROJECT_NAME):$(DOCKER_TAG)$(DOCKER_PLATFORM_TAG_SUFFIX)

.PHONY: docker_push_manifest
docker_push_manifest:
	# Push the manifest to the registry
	docker manifest push $(DOCKER_REGISTRY)/$(DOCKER_ORG)/$(PROJECT_NAME):$(DOCKER_TAG)

.PHONY: docker_delete_manifest
docker_delete_manifest:
	# Delete the manifest to the registry, ignore the error if manifest doesn't exist
	docker manifest rm $(DOCKER_REGISTRY)/$(DOCKER_ORG)/$(PROJECT_NAME):$(DOCKER_TAG) || true	
```

Adding targets related to handle the Docker image manifest which is useful to get useful info about it (i.e. layers, platforms and more).

**Review by ppatierno — CHANGES_REQUESTED** [review]

File: bin/docker/mqtt_bridge_run.sh [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+
+export JAVA_OPTS="${JAVA_OPTS} $(get_gc_opts)"
+
+# Disable FIPS if needed
+if [ "$FIPS_MODE" = "disabled" ]; then
+    export JAVA_OPTS="${JAVA_OPTS} -Dcom.redhat.fips=false"
+fi
+
+# starting MQTT Bridge with final configuration
+exec /usr/bin/tini -s -w -e 143 -- "${MYPATH}"/../mqtt_bridge_run.sh --config-file=/tmp/application.properties --mapping-rules=/tmp/topic-mapping-rules.json "$@"
```
[the remark this answers is not in the document]
Comment (antonio-pedro99) [reply bin/docker/mqtt_bridge_run.sh:30]: in HTTP Bridge, was it related to KAFKA_?
https://github.com/strimzi/strimzi-kafka-bridge/blob/36d2f2dcebaf9b15ff54c87d213b30ced2641541/bin/docker/kafka_bridge_run.sh#L77

I am not sure,

File: install/020-Deployment.yaml [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+      app: mqtt-bridge
+  template:
+    metadata:
+      labels:
+        app: mqtt-bridge
+    spec:
+      volumes:
+        - name: mqtt-bridge-config-volume
+          configMap:
+            name: mqtt-bridge-config
```
[the remark this answers is not in the document]
Comment (antonio-pedro99) [reply install/020-Deployment.yaml:20]: Does it implicate something? Or it is just preferable?

[the remark this answers is not in the document]
Comment (kyguy) [reply install/020-Deployment.yaml:20]: Just preferable

File: bin/docker/mqtt_bridge_run.sh (hunk 8 above)
[the remark this answers is not in the document]
Comment (kyguy) [reply bin/docker/mqtt_bridge_run.sh:30]: It just needs to be consistent with the mount path used in the `Deployment`, whatever mount path is used for the `ConfigMap`

[the remark this answers is not in the document]
Comment (ppatierno) [reply bin/docker/mqtt_bridge_run.sh:30]: In the HTTP bridge you have a `/tmp` because this is where the configuration is created if you see the upper part of the script. In our case, as Kyle says we need to use the mount path in the `Deployment`.

File: install/020-Deployment.yaml [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+      containers:
+        - name: mqtt-bridge
+          image: <your-image-here>
+          command: ["/opt/strimzi/bin/mqtt_bridge_run.sh"]
+          args: ["--config-file=/opt/strimzi/config/application.properties", "--mapping-rules=/opt/strimzi/config/topic-mapping-rules.json"]
+          ports:
+            - containerPort: 1883
+          volumeMounts:
+            - name: mqtt-bridge-config-volume
+              mountPath: /opt/strimzi/mqtt-bridge/
```
[the remark this answers is not in the document]
Comment (ppatierno) [reply install/020-Deployment.yaml:30]: Wdym?

File: install/030-Deployment.yaml [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+    matchLabels:
+      app: mqtt-bridge
+  template:
+    metadata:
+      labels:
+        app: mqtt-bridge
+    spec:
+      containers:
+        - name: mqtt-bridge
+          image: quay.io/antdev/mqtt-bridge:latest
```
Comment (antonio-pedro99) [inline install/030-Deployment.yaml:19]:
```suggestion
          image: <your-image-here>
```

File: bin/docker/mqtt_bridge_run.sh [hunk 12]
```diff
@@ -0,0 +1,24 @@
+#!/usr/bin/env bash
+set +x
```
Comment (ppatierno) [inline bin/docker/mqtt_bridge_run.sh:2]:
```suggestion
set +x

# Clean-up /tmp directory from files which might have remained from previous container restart
# We ignore any errors which might be caused by files injected by different agents which we do not have the rights to delete
rm -rfv /tmp/* || true
```

Would you mind to put this as well? We had issues in the past with `/tmp` folder growing out of our control.

File: bin/mqtt_bridge_run.sh [hunk 13]
```diff
@@ -10,4 +10,4 @@ then
       MQTT_BRIDGE_LOG4J_OPTS="-Dlog4j2.configurationFile=file:${MYPATH}/../config/log4j2.properties"
 fi
 
-exec java $MQTT_BRIDGE_LOG4J_OPTS -classpath "${MYPATH}/../libs/*" io.strimzi.kafka.bridge.mqtt.Main "$@"
+exec java "$MQTT_BRIDGE_LOG4J_OPTS" -classpath "${MYPATH}/../libs/*" io.strimzi.kafka.bridge.mqtt.Main "$@"
```
Comment (ppatierno) [inline bin/mqtt_bridge_run.sh:13]: why did you need to add double quotes here? Compared to the Strimzi HTTP bridge we didn't need them.
You are also missing to pass the `$JAVA_OPTS` env var at the beginning. It's not filled in this script but it could be filled by the docker related script calling this one.

File: bin/docker/mqtt_bridge_run.sh [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+  export JAVA_OPTS="-Xms${MAX_HEAP}m -Xmx${MAX_HEAP}m $JAVA_OPTS"
+fi
+
+# Disable FIPS if needed
+if [ "$FIPS_MODE" = "disabled" ]; then
+    export JAVA_OPTS="${JAVA_OPTS} -Dcom.redhat.fips=false"
+fi
+
+# starting MQTT Bridge with final configuration
+exec /usr/bin/tini -s -w -e 143 -- "${MYPATH}"/../mqtt_bridge_run.sh --config-file="${STRIMZI_HOME}"/config/application.properties --mapping-rules="${STRIMZI_HOME}"/config/topic-mapping-rules.json "$@"
```
Comment (ppatierno) [inline bin/docker/mqtt_bridge_run.sh:24]: why did you need double quotes on the config and mapping rules files parameters? Compared to the Strimzi HTTP bridge we didn't need them.

File: README.md [hunk 15]
```diff
@@ -8,6 +8,46 @@ This project provides a software component which acts as a bridge between [MQTT
...
+```shell
+bin/mqtt_bridge_run.sh --config-file <path/to/config/file> --mapping-rules <path/to/mapping/rules/file>
+```
+
+As a default, you can find the configuration and mapping rules files under the `config` directory.
+
+### Deploying on Kubernetes
+
+The MQTT Bridge is deployed using a Kubernetes `Deployment` , and it is configured using a `ConfigMap`.
+The ConfigMap contains the configuration file and the mapping rules file.
```
Comment (ppatierno) [inline README.md:42]:
```suggestion
The `ConfigMap` contains the configuration and the mapping rules files. 
```

File: README.md [hunk 16]
```diff
@@ -8,6 +8,46 @@ This project provides a software component which acts as a bridge between [MQTT
...
+
+As a default, you can find the configuration and mapping rules files under the `config` directory.
+
+### Deploying on Kubernetes
+
+The MQTT Bridge is deployed using a Kubernetes `Deployment` , and it is configured using a `ConfigMap`.
+The ConfigMap contains the configuration file and the mapping rules file. 
+The files under the `install` directory are used to deploy the MQTT Bridge on Kubernetes.
+
+To deploy the MQTT Bridge using the following command:
```
Comment (ppatierno) [inline README.md:45]:
```suggestion
To deploy the MQTT Bridge use the following command:
```

File: install/020-MqttBridgeConfigMap.yaml [hunk 17]
```diff
@@ -0,0 +1,17 @@
...
+data:
+  application.properties: |
+    #Bridge related settings
+    bridge.id=my-bridge
+
+    #MQTT server common
+    mqtt.server.host=localhost
+    mqtt.server.port=1883
+    #Apache Kafka common
+    kafka.bootstrap.servers=localhost:9092
```
Comment (ppatierno) [inline install/020-MqttBridgeConfigMap.yaml:16]: maybe we should have a placeholder here as well?  We don't know where the Kafka cluster is placed.

File: install/020-MqttBridgeConfigMap.yaml [hunk 18]
```diff
@@ -0,0 +1,17 @@
...
+  name: mqtt-bridge-config
+  labels:
+        app: mqtt-bridge
+data:
+  application.properties: |
+    #Bridge related settings
+    bridge.id=my-bridge
+
+    #MQTT server common
+    mqtt.server.host=localhost
```
Comment (ppatierno) [inline install/020-MqttBridgeConfigMap.yaml:13]:
```suggestion
    mqtt.server.host=0.0.0.0
```

Using 0.0.0.0 (all interfaces) by default allows the bridge to listen on all interfaces on the pod so accepting traffic from outside. With localhost it's not getting traffic unless the MQTT client is running on the same pod (which is usually not the case :-)).

**Review by ppatierno — CHANGES_REQUESTED** [review]

File: bin/docker/mqtt_bridge_run.sh (hunk 14 above)
Comment (antonio-pedro99) [reply bin/docker/mqtt_bridge_run.sh:24]: Does not double quotes prevent globbing and word splitting?

File: bin/mqtt_bridge_run.sh (hunk 13 above)
Comment (antonio-pedro99) [reply bin/mqtt_bridge_run.sh:13]: Does not double quotes prevent globbing and word splitting?

File: bin/docker/mqtt_bridge_run.sh (hunk 12 above)
Comment (antonio-pedro99) [reply bin/docker/mqtt_bridge_run.sh:2]: How come? We are not using this /tmp anywhere 

In the HTTP bridge, it was where the generated config were placed. 
Not the case with the MQTT Bridge, I guess so

File: install/020-MqttBridgeConfigMap.yaml (hunk 18 above)
Comment (antonio-pedro99) [reply install/020-MqttBridgeConfigMap.yaml:13]: gotcha!

File: bin/docker/mqtt_bridge_run.sh (hunk 12 above)
Comment (ppatierno) [reply bin/docker/mqtt_bridge_run.sh:2]: It was not the HTTP bridge there, it was Vert.x filling the tmp folder. We can't know everything about what our deps are doing.

Comment (antonio-pedro99) [reply bin/docker/mqtt_bridge_run.sh:2]: Oh, I see now.

File: install/020-MqttBridgeConfigMap.yaml (hunk 17 above)
Comment (ppatierno) [reply install/020-MqttBridgeConfigMap.yaml:16]: just use a placeholder for the address and leave 9092 as following port.

File: bin/mqtt_bridge_run.sh (hunk 13 above)
Comment (ppatierno) [reply bin/mqtt_bridge_run.sh:13]: but isn't it defined as a string already with double quotes?

File: bin/docker/mqtt_bridge_run.sh (hunk 14 above)
Comment (ppatierno) [reply bin/docker/mqtt_bridge_run.sh:24]: Thinking more these double quotes around `STRIMZI_HOME` makes sense.

File: install/020-MqttBridgeConfigMap.yaml (hunk 17 above)
Comment (ppatierno) [reply install/020-MqttBridgeConfigMap.yaml:16]: well no, actually it can be a comma separated list of Kafka brokers so even this one doesn't work fine.
Let's just use `<kafka-bootstrap-servers>` ... who uses Kafka knows what to put there.

Comment (antonio-pedro99) [reply install/020-MqttBridgeConfigMap.yaml:16]: Completely agree.

File: bin/mqtt_bridge_run.sh (hunk 13 above)
Comment (antonio-pedro99) [reply bin/mqtt_bridge_run.sh:13]: Yeah, but I keep getting warnings regarding it.

Comment (ppatierno) [reply bin/mqtt_bridge_run.sh:13]: which warnings? from where?

Comment (antonio-pedro99) [reply bin/mqtt_bridge_run.sh:13]: @ppatierno regarding the $JAVA_OPTS, was it not related to STRIMZI_ ? Because its first value was coming from `get_gc_opts` and inside this function we used `STRIMZI_GC_LOG_ENABLED`. 
Please clarify, maybe I got it wrong.

Comment (antonio-pedro99) [reply bin/mqtt_bridge_run.sh:13]: image
the warnings

Comment (ppatierno) [reply bin/mqtt_bridge_run.sh]: > @ppatierno regarding the $JAVA_OPTS, was it not related to STRIMZI_ ? Because its first value was coming from get_gc_opts and inside this function we used STRIMZI_GC_LOG_ENABLED.
Please clarify, maybe I got it wrong.

You are actually setting it in the bin/docker script for Max heap and FIPS but not using here. So you need it to be added on this command line :-)

> the warnings
I see but in the Strimzi HTTP bridge we had to disable this warning at some point (raised by shellcheck) because of some TLS related commands with kytool. Because we'll have TLS support one day in the future, I would follow the same pattern here.

File: install/020-MqttBridgeConfigMap.yaml [hunk 19]
```diff
@@ -0,0 +1,32 @@
+apiVersion: v1
```
Comment (ppatierno) [inline install/020-MqttBridgeConfigMap.yaml:1]: Just call it `020-ConfigMap.yaml` not need for the prefix. Otherwise you should have it for all the other YAMLs, which tbh is useless.

**Review by ppatierno — CHANGES_REQUESTED** [review]

File: install/020-ConfigMap.yaml [hunk 20]
```diff
@@ -0,0 +1,32 @@
...
+data:
+  application.properties: |
+    #Bridge related settings
+    bridge.id=my-bridge
+
+    #MQTT server common
+    mqtt.server.host=0.0.0.0
+    mqtt.server.port=1883
+    #Apache Kafka common
+    kafka.bootstrap.servers=0.0.0.0:9092
```
Comment (antonio-pedro99) [inline install/020-ConfigMap.yaml:16]:
```suggestion
    kafka.bootstrap.servers=<kafka-bootstrap-servers>
```
